### PR TITLE
Rails 3.2 to 4.0 article

### PR DESCRIPTION
**Pull request by mauro-oto**
Hey @etagwerker, @lubc, 

This PR is for the Rails 3.2 to Rails 4.0 article.

Please check it out, thanks!

### CODE REVIEW

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+covered some general tips to take into account for this migration. This article
+will try to go a bit more in depth. We will first go from 3.2 to 4.0, and then
+to 4.2. Depending on the complexity of your app, a Rails upgrade can take
+anywhere from one week for a single developer, to a few months for two developers.
+
+1. Ruby version
+2. Gems
+3. Config files (config/)
+4. Application code
+  a. Models (app/models/)
```
Comment (lubc) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: the markdown preview is not showing this in a new line

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+categories: ["rails", "upgrades"]
+author: "mauro-oto"
+---
+
+_This article is part of our Upgrade Rails series. To see more of them, [click here](https://www.ombulabs.com/blog/tags/upgrades)_.
+
+A [previous post](https://www.ombulabs.com/blog/rails/tips-for-upgrading-rails-3-2-to-4.html)
+covered some general tips to take into account for this migration. This article
+will try to go a bit more in depth. We will first go from 3.2 to 4.0, and then
+to 4.2. Depending on the complexity of your app, a Rails upgrade can take
```
Comment (lubc) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: what about `4.0` to `4.1`?

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+# before:
+has_many :posts, order: 'position'
+
+# after:
+has_many :posts, -> { order('position') }
+```
+
+- Protected Attributes are deprecated, but you can add the [gem for it](https://github.com/rails/protected_attributes).
+However, since the Rails core team dropped its support since Rails 5.0, you
+should begin migrating your models to Strong Parameters anyway.
```
Comment (lubc) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: You can add an example here

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 4]
```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: "The definitive guide to upgrading from Rails 3.2 to 4.0"
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto Please titleize this title. `The Definitive Guide To Upgrading From Rails 3.2 To 4.0` or `The Definitive Guide to Upgrading from Rails 3.2 to 4.0`

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+_This article is part of our Upgrade Rails series. To see more of them, [click here](https://www.ombulabs.com/blog/tags/upgrades)_.
+
+A [previous post](https://www.ombulabs.com/blog/rails/tips-for-upgrading-rails-3-2-to-4.html)
+covered some general tips to take into account for this migration. This article
+will try to go a bit more in depth. We will first go from 3.2 to 4.0, and then
+to 4.2. Depending on the complexity of your app, a Rails upgrade can take
+anywhere from one week for a single developer, to a few months for two developers.
+
+1. Ruby version
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto It'd be great if you can add a link to each item that later takes the visitor to the right section.

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+5. Tests
+6. Miscellaneous
+7. Next steps
+
+## 1. Ruby version
+
+Rails 3.2.x is the last version to support Ruby 1.8.7. If you're using Ruby 1.8.7,
+you'll need to upgrade to Ruby 1.9.3 or newer. The Ruby upgrade is out of the scope
+of this guide, but check out [this guide](http://www.darkridge.com/~jpr5/2012/10/03/ruby-1.8.7-1.9.3-migration), which
+is very complete and serves as a sort of checklist, and we've used it in the past.
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto Maybe separate this last sentence in two? 

```
The Ruby upgrade is not covered in this guide, but check out [this guide](http://www.darkridge.com/~jpr5/2012/10/03/ruby-1.8.7-1.9.3-migration). It's quite complete and it's a checklist we have used it in the past.
```

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+Instead of going through your currently bundled gems or `Gemfile.lock` manually,
+you get a report of the gems you need to upgrade.
+
+## 3. Config
+
+Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task),
+which you can use as a guideline as explained thoroughly in
+[this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html)
+to get rid of unnecessary code or monkey-patches in your config files and
+initializers, specially if your Rails 3 app was running on Rails 2.
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto This sentence is quite long and hard to follow. Try breaking it down into smaller sentences.

```
Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task). You can use this task as a guideline as explained thoroughly in [this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html). It will help you get rid of unnecessary code or monkey-patches in your config files and initializers, especially if your Rails 3 app was running on Rails 2.
```

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+[activerecord-deprecated_finders](https://github.com/rails/activerecord-deprecated_finders)
+
+- ActiveRecord scopes now need a lambda:
+
+```ruby
+# before:
+default_scope where(deleted_at: nil)
+
+# after:
+default_scope { where(deleted_at: nil) }
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto Maybe you can post a link to your previous article? 😊  https://www.ombulabs.com/blog/ruby/rails/best-practices/why-using-default-scope-is-a-bad-idea.html

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+# before:
+has_many :posts, order: 'position'
+
+# after:
+has_many :posts, -> { order('position') }
+```
+
+- Protected Attributes are deprecated, but you can add the [gem for it](https://github.com/rails/protected_attributes).
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto I think this sentence is more "SEO friendly" like this: 

```
Protected attributes' support is deprecated, but you can add the [protected_attributes](https://github.com/rails/protected_attributes) gem.
```

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+```
+
+- Rails 4.0 dropped support for plugins, so you'll need to replace them with gems,
+either by searching for the project on [RubyGems](https://rubygems.org)/[Github](https://github.com),
+or by moving the plugin to your `lib` directory and require it from somewhere
+within your Rails app.
+
+## 7. Next steps
+
+If you're running Rails 4.0, congratulations!
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto This is kind of weird. When they read this step, supposedly they followed your steps and you can write something like this: 

```
If you successfully followed all these steps by now you're running Rails 4.0. So congratulations!
```

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+Stay tuned for our Rails 4.0 to Rails 4.2 article, check out our [Rails performance articles](https://www.ombulabs.com/blog/tags/performance) to fine-tune
+your app, and feel free to tell us how your upgrade went.
```
Comment (etagwerker) [inline _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto Maybe the call to action can point to https://fastruby.io?

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown (hunk 2 above)
Comment (mauro-oto) [reply _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: :+1:

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown (hunk 1 above)
Comment (mauro-oto) [reply _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: Fixed, thanks

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown (hunk 3 above)
Comment (mauro-oto) [reply _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: Yep!

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown (hunk 4 above)
Comment (mauro-oto) [reply _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: Done!

File: _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown (hunk 5 above)
Comment (mauro-oto) [reply _posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown]: Yeah, done! :+1:

**Comment by mauro-oto** [comment]
@etagwerker @lubc just updated this post with your feedback, thanks!

**Comment by mauro-oto** [comment]
This PR is now ready for another review :+1:

File: _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 12]
```diff
@@ -0,0 +1,212 @@
+---
+layout: post
+title: "The Definitive Guide to Upgrading from Rails 3.2 to 4.0"
+date: 2017-10-25 16:48:00
```
Comment (etagwerker) [inline _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto Please update the date.

File: _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+default_scope { where(deleted_at: nil) }
+
+# before:
+has_many :posts, order: 'position'
+
+# after:
+has_many :posts, -> { order('position') }
+```
+
+(Friendly reminder: beware when using [default_scope](https://www.ombulabs.com/blog/ruby/rails/best-practices/why-using-default-scope-is-a-bad-idea.html))
```
Comment (etagwerker) [inline _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown]: 👍

File: _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+- Protected attributes is deprecated, but you can still add the [protected_attributes](https://github.com/rails/protected_attributes) gem.
+However, since the Rails core team dropped its support since Rails 5.0, you
+should begin migrating your models to Strong Parameters anyway.
```
Comment (etagwerker) [inline _posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown]: @mauro-oto please add a link to http://guides.rubyonrails.org/action_controller_overview.html#strong-parameters

**Comment by mauro-oto** [comment]
@etagwerker thanks! just updated it.

**Comment by etagwerker** [comment]
@mauro-oto cool, thanks! could you solve the conflicts?
